File: src/prototype/web/src/state/pass_attempt.ts
// PURE — pass-attempt state per docs/design/input_system_defense_v1.md §B.7.
//
// M1 simplification: we skip the 5-second animation phase and produce one
// of three terminal outcomes on commit:
//   - ineligible     → commit silently rejected (no state change, no event)
//   - counter hits   → PASS_FAILED (attacker triangle-counters during the
//                      5s window; represented as counterConfirmed=true)
//   - otherwise      → PASS_SUCCEEDED after passWindowMs elapses
//
// The pass attempt lives on GameState so stepSimulation can tick its
// timer. Caller supplies the raw PASS_COMMIT intent; we gate it here.

import type { ActorState, GuardState } from "./game_state.js";

export type PassAttemptState =
  | { kind: "IDLE" }
  | { kind: "IN_PROGRESS"; startedMs: number };

export const INITIAL_PASS_ATTEMPT: PassAttemptState = Object.freeze({ kind: "IDLE" });

export const PASS_TIMING = Object.freeze({
  windowMs: 5000,
});

// §B.7.1 — eligibility predicates.
export function isPassEligible(params: Readonly<{
  bottom: ActorState;
  top: ActorState;
  defenderStamina: number;
  leftBasePressure: number;
  rightBasePressure: number;
  leftBaseZone: string | null;
  rightBaseZone: string | null;
  rsY: number;
  guard: GuardState;
}>): boolean {
  if (params.guard !== "CLOSED") return false;
  // One foot must be UNLOCKED (closed guard not completely sealing).
  const oneFootUnlocked =
    params.bottom.leftFoot.state === "UNLOCKED" ||
    params.bottom.rightFoot.state === "UNLOCKED";
  if (!oneFootUnlocked) return false;

  if (params.defenderStamina < 0.2) return false;

  // Both hands pressing ≥ 0.5 on BICEP_* or KNEE_*.
  const controlZone = (z: string | null) =>
    z === "BICEP_L" || z === "BICEP_R" || z === "KNEE_L" || z === "KNEE_R";
  if (
    !controlZone(params.leftBaseZone) ||
    !controlZone(params.rightBaseZone) ||
    params.leftBasePressure < 0.5 ||
    params.rightBasePressure < 0.5
  ) {
    return false;
  }

  // RS pointing downward (rsY ≤ 0 with some magnitude) per §B.7.1.
  if (params.rsY > -0.3) return false;

  return true;
}

export type PassTickInput = Readonly<{
  nowMs: number;
  commitRequested: boolean;
  eligibleNow: boolean;
  // If the attacker pulls off a TRIANGLE during the 5s window the pass
  // fails outright. In Stage 1 we receive this as a signal from the
  // judgment-window layer.
  attackerTriangleConfirmedThisTick: boolean;
}>;

export type PassTickEvent =
  | { kind: "PASS_STARTED" }
  | { kind: "PASS_FAILED" }
  | { kind: "PASS_SUCCEEDED" };

export function tickPassAttempt(
  prev: PassAttemptState,
  inp: PassTickInput,
  timing = PASS_TIMING,
): { next: PassAttemptState; events: readonly PassTickEvent[] } {
  const events: PassTickEvent[] = [];

  if (prev.kind === "IDLE") {
    if (inp.commitRequested && inp.eligibleNow) {
      events.push({ kind: "PASS_STARTED" });
      return {
        next: Object.freeze({ kind: "IN_PROGRESS" as const, startedMs: inp.nowMs }),
        events: Object.freeze(events),
      };
    }
    return { next: prev, events: Object.freeze(events) };
  }

  // IN_PROGRESS: attacker's triangle aborts the pass immediately.
  if (inp.attackerTriangleConfirmedThisTick) {
    events.push({ kind: "PASS_FAILED" });
    return { next: INITIAL_PASS_ATTEMPT, events: Object.freeze(events) };
  }

  if (inp.nowMs - prev.startedMs >= timing.windowMs) {
    events.push({ kind: "PASS_SUCCEEDED" });
    return { next: INITIAL_PASS_ATTEMPT, events: Object.freeze(events) };
  }

  return { next: prev, events: Object.freeze(events) };
}

File: src/prototype/unity/Assets/BJJSimulator/Runtime/State/PassAttempt.cs
// Ported 1:1 from src/prototype/web/src/state/pass_attempt.ts.
// PURE — pass-attempt state per docs/design/input_system_defense_v1.md §B.7.

using System.Collections.Generic;

namespace BJJSimulator
{
    // -------------------------------------------------------------------------
    // State
    // -------------------------------------------------------------------------

    public enum PassAttemptKind { Idle, InProgress }

    public struct PassAttemptState
    {
        public PassAttemptKind Kind;
        public long            StartedMs; // valid only when Kind == InProgress

        public static readonly PassAttemptState Idle = new PassAttemptState
        {
            Kind      = PassAttemptKind.Idle,
            StartedMs = BJJConst.SentinelTimeMs,
        };
    }

    // -------------------------------------------------------------------------
    // Timing
    // -------------------------------------------------------------------------

    public struct PassTiming
    {
        public int WindowMs; // 5000

        public static readonly PassTiming Default = new PassTiming { WindowMs = 5000 };
    }

    // -------------------------------------------------------------------------
    // Eligibility inputs (§B.7.1)
    // -------------------------------------------------------------------------

    public struct PassEligibilityParams
    {
        public ActorState Bottom;
        public ActorState Top;
        public float      DefenderStamina;
        public float      LeftBasePressure;
        public float      RightBasePressure;
        public BaseZone   LeftBaseZone;
        public BaseZone   RightBaseZone;
        public float      RsY;
        public GuardState Guard;
    }

    // -------------------------------------------------------------------------
    // Tick input
    // -------------------------------------------------------------------------

    public struct PassTickInput
    {
        public long NowMs;
        public bool CommitRequested;
        public bool EligibleNow;
        // If the attacker pulls off a TRIANGLE during the 5s window, the pass fails.
        public bool AttackerTriangleConfirmedThisTick;
    }

    // -------------------------------------------------------------------------
    // Events
    // -------------------------------------------------------------------------

    public enum PassEventKind { PassStarted, PassFailed, PassSucceeded }

    public struct PassTickEvent
    {
        public PassEventKind Kind;
    }

    // -------------------------------------------------------------------------
    // Pure operations
    // -------------------------------------------------------------------------

    public static class PassAttemptOps
    {
        // §B.7.1 — eligibility check.
        public static bool IsPassEligible(PassEligibilityParams p)
        {
            if (p.Guard != GuardState.Closed) return false;

            bool oneFootUnlocked =
                p.Bottom.LeftFoot.State == FootState.Unlocked ||
                p.Bottom.RightFoot.State == FootState.Unlocked;
            if (!oneFootUnlocked) return false;

            if (p.DefenderStamina < 0.2f) return false;

            if (!IsControlZone(p.LeftBaseZone) || !IsControlZone(p.RightBaseZone)) return false;
            if (p.LeftBasePressure < 0.5f || p.RightBasePressure < 0.5f) return false;

            if (p.RsY > -0.3f) return false;

            return true;
        }

        public static PassAttemptState Tick(
            PassAttemptState prev,
            PassTickInput inp,
            List<PassTickEvent> events,
            PassTiming timing = default)
        {
            if (timing.WindowMs == 0) timing = PassTiming.Default;

            if (prev.Kind == PassAttemptKind.Idle)
            {
                if (inp.CommitRequested && inp.EligibleNow)
                {
                    events.Add(new PassTickEvent { Kind = PassEventKind.PassStarted });
                    return new PassAttemptState { Kind = PassAttemptKind.InProgress, StartedMs = inp.NowMs };
                }
                return prev;
            }

            // IN_PROGRESS
            if (inp.AttackerTriangleConfirmedThisTick)
            {
                events.Add(new PassTickEvent { Kind = PassEventKind.PassFailed });
                return PassAttemptState.Idle;
            }

            if (inp.NowMs - prev.StartedMs >= timing.WindowMs)
            {
                events.Add(new PassTickEvent { Kind = PassEventKind.PassSucceeded });
                return PassAttemptState.Idle;
            }

            return prev;
        }

        // §B.7.1 control zones: BICEP_* or KNEE_*
        static bool IsControlZone(BaseZone z) =>
            z == BaseZone.BicepL || z == BaseZone.BicepR ||
            z == BaseZone.KneeL  || z == BaseZone.KneeR;
    }
}
